rclpy/rclpy/context.py

```diff
@@ -123,23 +140,30 @@ def try_shutdown(self):
                 self._call_on_shutdown_callbacks()
                 self._logging_fini()
 
-    def _remove_callback(self, weak_method):
+    def _remove_callback(self, weak_method: WeakMethod[MethodType]) -> None:
         self._callbacks.remove(weak_method)
 
-    def on_shutdown(self, callback: Callable[[], None]):
+    def on_shutdown(self, callback: Callable[[], None]) -> None:
         """Add a callback to be called on shutdown."""
         if not callable(callback):
             raise TypeError('callback should be a callable, got {}', type(callback))
+        if self.__context is None:
```

Comment:
Oh I see, before we were holding `self.__context`, but if that was None then the `with` statement failed. Would you be willing to wrap this new block in `with self._lock:`?